### PR TITLE
Sync with GSL RRFS_dev1 workflow and app

### PR DESCRIPTION
**Pull request by hu5970**
Sync with:
ufs_srweather_app RRFS_dev1 code changes from hash b6f1d9f85a (tag rrfs_a.v0.7.1) to d922123a769 (Nov 8, 2023)
Sync NOAA-GSL feature/RRFS_dev regional workflow hash 9e099e719 (Oct 18,2023) to d3df76f (Nov 13,2023) code changes

The retro tests give identical results for deterministic and EnKF DA runs comparing with the results from NOAA-GSL regional workflow and APP for their head of RRFS_dev1 branch (Nov 13,2023). 

-

## TESTS CONDUCTED: 

- On machines/platforms:

- [ ] WCOSS2
- [x ] Hera
- [ ] Orion
- [ ] Jet

### CODE REVIEW

**Comment by hu5970** [comment]
Hi All, This PR is to bring the rest of the NOAA-GSL workflow and APP changes to the rrfs-workflow. This will be the base to run v0.7.5 with configuration changes. We will try to freeze the NOAA GSL workflow.

File: modulefiles/tasks/wcoss2/make_ics.local.lua
Comment (chan-hoo) [inline modulefiles/tasks/wcoss2/make_ics.local.lua:22]: @hu5970, what is the difference between BLENDINGPYTHON and general(?) python? Does it use any specific packages that are not installed in python on wcoss2?

Comment (delippi) [reply modulefiles/tasks/wcoss2/make_ics.local.lua:22]: @chan-hoo, sorry if there is confusion. This is just the version of (general) python that I was using on WCOSS2 for the blending function. It does not use any other specific packages that were not already installed in python on WCOSS2.

Comment (chan-hoo) [reply modulefiles/tasks/wcoss2/make_ics.local.lua:22]: @delippi, I got it. Thanks. If so, @MatthewPyle-NOAA, we can merge this PR as it is. However, we'll need to remove this 'BLENDINGPYTHON' in the near future, and set up a new python environment including additional python packages such as netCDF4 and numpy on other machines (Hera, Jet, Orion) because the current "workflow_tool" env may not have them as far as I know. I'll open an issue for this.

Comment (MatthewPyle-NOAA) [reply modulefiles/tasks/wcoss2/make_ics.local.lua:22]: @chan-hoo Any objection to working on this PR before your PR to clean up the home directory (and other things from the AQM review)?

Comment (chan-hoo) [reply modulefiles/tasks/wcoss2/make_ics.local.lua:22]: Nope. :)

File: scripts/exrrfs_run_post.sh
Comment (MatthewPyle-NOAA) [inline scripts/exrrfs_run_post.sh:1]: I need to double check the discussion about this one in the RRFS_dev1 merge.  An actual HRRR 3km output file matches the specs of the old definition

**Review by MatthewPyle-NOAA — COMMENTED** [review]
A couple of minor comments - want to double check one detail before approving.

File: scripts/exrrfs_run_post.sh
Comment (hu5970) [reply scripts/exrrfs_run_post.sh:1]: Eric James made a PR to change this number yesterday.  Please double check it.

File: ush/set_extrn_mdl_params.sh
Comment (hu5970) [reply ush/set_extrn_mdl_params.sh:1]: Agree. In general, we need to refactor how to setup outside model resources for lbcs an ics. The current method is very messy.

File: modulefiles/tasks/wcoss2/make_ics.local.lua
Comment (hu5970) [reply modulefiles/tasks/wcoss2/make_ics.local.lua:22]: @chan-hoo @delippi Thanks for working on python environment  setup. We do have problem to use python in different machines. There are several other tasks also using python (smoke/dust preprocess and fvcom preprocess). Need a good python environment works for those tasks on Hera/Jet/WCOSS2.

**Comment by hu5970** [comment]
@chan-hoo One issue I found in testing rrfs-workflow on Hera and Jet is "err_chk" after each executable does not work. So the task will not report as a failure task if executable does not run. The error message is:

`/scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/intel-2022.1.2/prod_util/2.0.14/bin/err_exit: line 62: ECF_NAME: parameter null or not set`

Is there any way to fix this?

Thanks,
Ming

**Comment by chan-hoo** [comment]
@hu5970, yes, I agree. I found it too (https://github.com/NOAA-EMC/rrfs-workflow/issues/86). I didn't have time to fix this. I think we should add the function call `print_err_msg_exit` when err != 0 only for rocoto. This might be because NCO doesn't want this kind of failure (DEAD). In my test, `err_exit` didn't work correctly. I think `print_err_msg_exit` is the best option at this moment.

**Comment by MatthewPyle-NOAA** [comment]
@chan-hoo @hu5970 To properly use err_exit, we may need to exporting more variables.   I think setting SENDECF=NO should help in particular.

USAGE:  To use this script one must export the following variables to the
 script:` jobid, SENDECF, pgm, pgmout, DATA`

**Comment by MatthewPyle-NOAA** [comment]
@hu5970 Good point.  Unfortunately most/all production utilities are very WCOSS2-centric.  Changing them isn't a trivial thing, so we need to think about ways to side step them on non-WCOSS2 machines.  Not sure if this idea is viable, but was wondering if define err_chk and err_exit at a high level in workflow to use items that work for that machine (and avoid the prod_util versions).  Also a comment from Rahul in [PR 57](https://github.com/NOAA-EMC/rrfs-workflow/pull/57) that is relevant here:

"err_chk and err_exit are utilities in prod_util. If you/anyone has updates to these that can make them portable (machine and scheduler), a wider community would benefit. Please consider contributing to [NCEPLibs-prod_util](https://github.com/NOAA-EMC/NCEPLIBS-prod_util)"

**Comment by MatthewPyle-NOAA** [comment]
@hu5970 Can you update your branch?  I'd like to get this merged within the next day.  Thanks!

**Comment by hu5970** [comment]
@MatthewPyle-NOAA What update do I need to do for my branch? I did not see comment that requires cod change.

On "err_exit", Guoqing will make a new prod_util module on JET/Hera and make "prod_util" work for Jet and Hera by switching to new prod_util. This is temporal solution for running system on Jet/Hera. Let's figure out how to port "err_exit" on WCOSS2, Hera, Jet in the future.

**Comment by guoqing-noaa** [comment]
FYI, I created a PR https://github.com/NOAA-EMC/NCEPLIBS-prod_util/pull/69 trying to address the `err_exit` issue.

Don't know how long it will take for the PR to get merged and implemented in different platforms.

I would suggest we use a temporary prod_util location for the workflow on Hera/Jet so that end users don't need to deal with a weird situation where tasks do not quit on errors. 

Once a new prod_util is ready to use, we can change to this new version.

**Comment by guoqing-noaa** [comment]
> @hu5970 Can you update your branch? I'd like to get this merged within the next day. Thanks!

@hu5970 I think @MatthewPyle-NOAA means that there are new commits in the latest dev-sci branch and we need to merge those changes to this PR before a merge.

**Comment by guoqing-noaa** [comment]
@hu5970 Could you incorporate my changes on `build_hera_intel.lua` and `build_jet_intel.lua` from this PR https://github.com/hu5970/rrfs-workflow/pull/1? Thanks!

**Comment by hu5970** [comment]
@MatthewPyle-NOAA @chan-hoo  I synced the PR with the latest dev-sci branch. Also, Guoqing has a temporal fix for err_exit on Jet/Hera. It should not impact wcoss2.

**Review by MatthewPyle-NOAA — APPROVED** [review]
Giving it a thumbs up - don't want to stand in the way of v0.7.5 work.

**Comment by MatthewPyle-NOAA** [comment]
@hu5970 I'm read to merge this if that sounds good to you.  Are you okay with a squash and merge?

**Comment by BenjaminBlake-NOAA** [comment]
@MatthewPyle-NOAA  Somehow I never saw this PR until just now.   I'm not sure why Github didn't notify me about it.  I would have waited to open my latest fire weather PR until this one was merged, so sorry about that.

**Comment by hu5970** [comment]
@MatthewPyle-NOAA Yes, please make it with a squash and merge

**Comment by MatthewPyle-NOAA** [comment]
Decided to do a non-DA Engineering test on WCOSS, and had a failure in make_grid with a complaint about a non-existent fvcom_mask file when executing link_fix.sh

**Comment by hu5970** [comment]
@MatthewPyle-NOAA Could you check:
rrfs-workflow/fix/lam/RRFS_NA_3km to see if you have the following link:
C3463_fvcom_mask.nc -> ../../.agent/lam/RRFS_NA_3km_C3463_Lake_fracSV/C3463_fvcom_mask.nc

rrfs-workflow/fix/lam/RRFS_CONUS_3km to see if you have the following link:
C3359_fvcom_mask.nc -> ../../.agent/lam/RRFS_CONUS_3km_C3359_Lake_fracSV/C3359_fvcom_mask.nc

**Comment by MatthewPyle-NOAA** [comment]
@hu5970 Yes, have those, but the engineering tests are typically with a 25 km CONUS domain.  I'll just create an issue for this to be fixed.

**Comment by BenjaminBlake-NOAA** [comment]
@MatthewPyle-NOAA  I fixed the issue with the fvcom file causing an error in make_grid in my latest PR for the fire weather grid.

**Comment by MatthewPyle-NOAA** [comment]
Thanks @BenjaminBlake-NOAA

**Comment by MatthewPyle-NOAA** [comment]
Merging now as noted problem with make_grid is ready to be fixed in an upcoming PR.

**Pull request merged**
